specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+
+
+![](https://hackmd.io/_uploads/HkAodIuFc.png)
+
+### Checkpoints data structure
+Checkpoints are always identified though the [Content Identifier (CID)](https://github.com/multiformats/cid) of their `Data` (i.e. the payload of the checkpoint), and they can optionally include the corresponding signature from validators in the subnet chain (this can be the signature of an individual miner, a multi-signature, or a threshold signature, depending on the `SA` policy). The signature is never used for the computation of the CID of the checkpoint.
+
+> Checkpoints Data Structure
+```go
+
+// Checkpoints wrap the checkpoint data and
+// a field dedicated for the arbitrary signing policy
+// used by the subnet.
+type Checkpoint struct {
+    Data CheckpointData
+    Sig []byte
+}
+
+// Data included in checkpoints
+type CheckpointData struct {
+    // Source subnet of the checkpoint.
+    Source     address.SubnetID
+    // Proof of the state of the subnet. In the
+    // reference implementation this is the tipset
+    // for the epoch in the checkpoint, but this could
+    // be a ZKProof or any other proof chosen by the subnet.
+    Proof     []byte
+    // Epoch of the checkpoint. Determined by the
+    // subnet's CheckPeriod
+    Epoch     abi.ChainEpoch
+    // Cid of the previous checkpoint committed for the subnet.
+    PrevCheck cid.Cid
+    // Array with the aggregation of all the checkpoints committed
+    // from the subnet's children.
+    Children  []ChildCheck
+    // Array with cross-message metadata of messages being propagated
+    // in the current checkpoint.
+    CrossMsgs []CrossMsgMeta
+}
+
+// Metadata pointing to the list of messages being
+// propagated in the checkpoint.
+type CrossMsgMeta struct {
+    // Source of the cross-messages included in
+    // this metadata packagle
+    From    address.SubnetID
+    // Destination of the cross-messages included.
+    To      address.SubnetID
+    // CID of the aggregation of all the messages being
+    // propagated. This CID is used to uniquely identify
+    // this package of messages.
+    MsgsCid cid.Cid
+    // Nonce of the crossMsgMeta. It is used for partial
+    // ordering and to prevent replay attacks.
+    Nonce   uint64
+    // Aggregation of all the native tokens being transacted
+    // in the included messages.
+    Value   abi.TokenAmount
+}
+
+// Package with all checkpoints committed in
+// the checkpoint window from children.
+type ChildChecks struct {
+    // Source of the checkpoints. ID of the child.
+    Source address.SubnetID
+    // List of cid of the checkpoints committed by
+    // the child in this checkpoint window.
+    Checks []cid.Cid
+}
+
+```
+Checkpoints include in their payload the following data:
```

Comment:
This seems a little duplicative of the structure description above. I realise it's more detailed, but another point of potential failure (someone updates structure above, forgets to propagate here).

Reply:
Fixed.